File: app/api/routes/deep_search.py
"""Deep search API endpoints."""

from fastapi import APIRouter, Depends, HTTPException
from sqlalchemy.ext.asyncio import AsyncSession

from app.api.dependencies import get_db
from app.core import get_logger
from app.deep_search.graph import run_deep_search
from app.models.schemas import (
    CollectedInfoResponse,
    DeepSearchRequest,
    DeepSearchResponse,
    ToolCallInfo,
)

router = APIRouter()
logger = get_logger(__name__)


@router.post("/deepSearch", response_model=DeepSearchResponse)
async def execute_deep_search(
    request: DeepSearchRequest,
    db: AsyncSession = Depends(get_db),
) -> DeepSearchResponse:
    """Execute deep search for an article.

    This endpoint triggers a ReAct-style workflow to collect background
    information and generate a comprehensive tracking report.

    Args:
        request: Deep search request with article_id and max_iterations
        db: Database session

    Returns:
        Deep search response with final report and tool history
    """
    logger.info(
        "Deep search requested",
        article_id=request.article_id,
        max_iterations=request.max_iterations,
    )

    try:
        # Validate UUID format
        import uuid
        try:
            uuid.UUID(request.article_id)
        except ValueError:
            raise HTTPException(
                status_code=400,
                detail="Invalid article_id format. Must be a valid UUID.",
            )

        # Execute deep search
        state = await run_deep_search(
            session=db,
            article_id=request.article_id,
            max_iterations=request.max_iterations,
        )

        # Check if article was found
        if not state.get("article"):
            raise HTTPException(
                status_code=404,
                detail="Article not found",
            )

        # Build response
        article = state.get("article", {})
        tool_calls = [
            ToolCallInfo(
                tool_name=tc["tool_name"],
                tool_input=tc["tool_input"],
                tool_output=tc["tool_output"],
                iteration=tc["iteration"],
            )
            for tc in state.get("tool_history", [])
        ]

        collected_info = [
            CollectedInfoResponse(
                source=ci["source"],
                content=ci["content"],
                relevance=ci["relevance"],
                metadata=ci["metadata"],
            )
            for ci in state.get("collected_info", [])
        ]

        return DeepSearchResponse(
            article_id=request.article_id,
            article_title=article.get("title", ""),
            final_report=state.get("final_report", ""),
            tools_used=tool_calls,
            collected_info=collected_info,
            iterations=state.get("current_iteration", 0),
            is_complete=state.get("is_complete", False),
            errors=state.get("errors", []),
        )

    except HTTPException:
        raise
    except Exception as e:
        logger.error("Deep search failed", error=str(e))
        raise HTTPException(status_code=500, detail=f"Deep search failed: {str(e)}")

File: ios-app/TechNewsIOSApp/Models/DeepSearch.swift
import Foundation

struct DeepSearchRequest: Encodable {
    let articleID: String
    let maxIterations: Int

    enum CodingKeys: String, CodingKey {
        case articleID = "article_id"
        case maxIterations = "max_iterations"
    }
}

struct DeepSearchResult: Decodable, Equatable {
    let articleID: String
    let articleTitle: String
    let finalReport: String
    let toolsUsed: [ToolCallInfo]
    let collectedInfo: [CollectedInfo]
    let iterations: Int
    let isComplete: Bool
    let errors: [DeepSearchError]

    enum CodingKeys: String, CodingKey {
        case articleID = "article_id"
        case articleTitle = "article_title"
        case finalReport = "final_report"
        case toolsUsed = "tools_used"
        case collectedInfo = "collected_info"
        case iterations
        case isComplete = "is_complete"
        case errors
    }
}

struct ToolCallInfo: Decodable, Equatable, Hashable {
    let toolName: String
    let toolInput: [String: StringCodable]
    let toolOutput: String
    let iteration: Int

    enum CodingKeys: String, CodingKey {
        case toolName = "tool_name"
        case toolInput = "tool_input"
        case toolOutput = "tool_output"
        case iteration
    }
}

struct CollectedInfo: Decodable, Equatable, Hashable {
    let source: String
    let content: String
    let relevance: String
}

struct DeepSearchError: Decodable, Equatable, Hashable {
    let phase: String
    let message: String
}

enum StringCodable: Codable, Equatable, Hashable {
    case string(String)
    case int(Int)
    case double(Double)
    case bool(Bool)

    init(from decoder: Decoder) throws {
        let container = try decoder.singleValueContainer()
        if let value = try? container.decode(String.self) {
            self = .string(value)
        } else if let value = try? container.decode(Int.self) {
            self = .int(value)
        } else if let value = try? container.decode(Double.self) {
            self = .double(value)
        } else if let value = try? container.decode(Bool.self) {
            self = .bool(value)
        } else {
            self = .string("")
        }
    }

    func encode(to encoder: Encoder) throws {
        var container = encoder.singleValueContainer()
        switch self {
        case .string(let value):
            try container.encode(value)
        case .int(let value):
            try container.encode(value)
        case .double(let value):
            try container.encode(value)
        case .bool(let value):
            try container.encode(value)
        }
    }
}
